framework/auth/views.py

```diff
@@ -51,88 +57,134 @@ def reset_password(auth, **kwargs):
+            # check if the user exists
+            user_obj = get_user(email=email)
+            if user_obj:
+                # check forgot_password rate limit
+                if throttle_period_expired(user_obj.email_last_sent, settings.SEND_EMAIL_THROTTLE):
+                    # new random verification key, allows OSF to check whether the reset_password request is valid,
+                    # this verification key is used twice, one for GET reset_password and one for POST reset_password
+                    # and it will be destroyed when POST reset_password succeeds
+                    user_obj.verification_key = generate_verification_key()
+                    user_obj.email_last_sent = datetime.datetime.utcnow()
+                    user_obj.save()
+                    reset_link = furl.urljoin(
+                        settings.DOMAIN,
+                        web_url_for(
+                            'reset_password',
+                            verification_key=user_obj.verification_key
+                        )
                     )
-                )
-                mails.send_mail(
-                    to_addr=email,
-                    mail=mails.FORGOT_PASSWORD,
-                    reset_link=reset_link
-                )
-                status.push_status_message(status_message, kind='success', trust=False)
+                    mails.send_mail(
+                        to_addr=email,
+                        mail=mails.FORGOT_PASSWORD,
+                        reset_link=reset_link
+                    )
+                    status.push_status_message(status_message, kind='success', trust=False)
+                else:
+                    status.push_status_message('You have recently requested to change your password. Please wait a '
+                                               'little while before trying again.', kind='error', trust=False)
             else:
-                status.push_status_message('You have recently requested to change your password. Please wait a little '
-                                           'while before trying again.',
-                                           kind='error',
-                                           trust=False)
+                status.push_status_message(status_message, kind='success', trust=False)
+            # TODO: disable form upon success
         else:
-            status.push_status_message(status_message, kind='success', trust=False)
-    forms.push_errors_to_status(form.errors)
-    return auth_login(forgot_password_form=form)
+            forms.push_errors_to_status(form.errors)
+            # Don't go anywhere
+
+    return {}
 
 
 @collect_auth
-def forgot_password_get(auth, *args, **kwargs):
-    """Return forgot password page upon.
+def auth_login(auth):
     """
-    if auth.logged_in:
-        return redirect(web_url_for('dashboard'))
-    return {}
+    This view serves as the entry point for OSF login and campaign login.
```

Comment:
Note for CR: this comment is a 'reverse-engineering' of how the following code work.